contracts/XYZSwapPair.sol

```diff
@@ -90,71 +123,92 @@ contract XYZSwapPair is IXYZSwapPair, ERC20Permit, ReentrancyGuard, VolumeTrendR
+        if (feeOn) kLast = getK(isAmpPool, _data); // data are up-to-date
         emit Burn(msg.sender, amount0, amount1, to);
     }
 
     /// @dev this low-level function should be called from a contract
-    ///             which performs important safety checks
+    /// @dev which performs important safety checks
     function swap(
         uint256 amount0Out,
         uint256 amount1Out,
         address to,
-        bytes calldata data
+        bytes calldata callbackData
     ) external override nonReentrant {
         require(amount0Out > 0 || amount1Out > 0, "XYZSwap: INSUFFICIENT_OUTPUT_AMOUNT");
-        (uint256 _reserve0, uint256 _reserve1, ) = getReserves(); // gas savings
+        (bool isAmpPool, ReserveData memory data) = getReservesData(); // gas savings
         require(
-            amount0Out < _reserve0 && amount1Out < _reserve1,
+            amount0Out < data.reserve0 && amount1Out < data.reserve1,
             "XYZSwap: INSUFFICIENT_LIQUIDITY"
         );
 
-        uint256 balance0;
-        uint256 balance1;
+        ReserveData memory newData;
         {
             // scope for _token{0,1}, avoids stack too deep errors
             IERC20 _token0 = token0;
             IERC20 _token1 = token1;
             require(to != address(_token0) && to != address(_token1), "XYZSwap: INVALID_TO");
             if (amount0Out > 0) _token0.safeTransfer(to, amount0Out); // optimistically transfer tokens
             if (amount1Out > 0) _token1.safeTransfer(to, amount1Out); // optimistically transfer tokens
-            if (data.length > 0)
-                IXYZSwapCallee(to).xyzSwapCall(msg.sender, amount0Out, amount1Out, data);
-            balance0 = _token0.balanceOf(address(this));
-            balance1 = _token1.balanceOf(address(this));
+            if (callbackData.length > 0)
+                IXYZSwapCallee(to).xyzSwapCall(msg.sender, amount0Out, amount1Out, callbackData);
+            newData.reserve0 = _token0.balanceOf(address(this));
+            newData.reserve1 = _token1.balanceOf(address(this));
+            if (isAmpPool) {
+                newData.vReserve0 = data.vReserve0.add(newData.reserve0).sub(data.reserve0);
+                newData.vReserve1 = data.vReserve1.add(newData.reserve1).sub(data.reserve1);
+            }
         }
-        uint256 amount0In = balance0 > _reserve0 - amount0Out
-            ? balance0 - (_reserve0 - amount0Out)
+        uint256 amount0In = newData.reserve0 > data.reserve0 - amount0Out
+            ? newData.reserve0 - (data.reserve0 - amount0Out)
             : 0;
-        uint256 amount1In = balance1 > _reserve1 - amount1Out
-            ? balance1 - (_reserve1 - amount1Out)
+        uint256 amount1In = newData.reserve1 > data.reserve1 - amount1Out
+            ? newData.reserve1 - (data.reserve1 - amount1Out)
             : 0;
         require(amount0In > 0 || amount1In > 0, "XYZSwap: INSUFFICIENT_INPUT_AMOUNT");
+        uint256 feeInPrecision = verifyBalanceAndUpdateEma(
```

Comment:
should we do this instead?
`uint256 feeInPrecision = isAmpPool ? verifyBalanceAndUpdateEma() : verifyBalanceAndUpdateEma()`

Reply:
I try, actually it looks uglier

Reply:
well I think we could save some gas by checking value of `isAmpPool1` only once, but maybe just a few gas